File: src/ios/Classes/BoatPlugin.swift
import AVFoundation
import Flutter
import UIKit

public class BoatPlugin: NSObject, FlutterPlugin {

    private var eventsSink: FlutterEventSink?
    private var captureSink: FlutterEventSink?

    private var sessionManager: AudioSessionManager?
    private var captureEngine: AudioCaptureEngine?
    private var playbackEngine: AudioPlaybackEngine?
    private var routeManager: AudioRouteManager?
    private var pipeline: PipelineRunner?
    private var publisher: FramePublisher?

    private let permissionManager = PermissionManager()
    private var currentState = "idle"
    private var startTimeMs: Double = 0
    private var lastConfig: [String: Any]?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methods = FlutterMethodChannel(name: "com.circuids.boat/methods",
                                           binaryMessenger: registrar.messenger())
        let events = FlutterEventChannel(name: "com.circuids.boat/events",
                                         binaryMessenger: registrar.messenger())
        let capture = FlutterEventChannel(name: "com.circuids.boat/capture",
                                          binaryMessenger: registrar.messenger())

        let instance = BoatPlugin()
        registrar.addMethodCallDelegate(instance, channel: methods)

        events.setStreamHandler(instance)
        capture.setStreamHandler(CaptureStreamHandler { instance.captureSink = $0 })
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            handleStart(call, result)
        case "stop":
            teardown()
            emitState("idle")
            result(nil)
        case "pause":
            handlePause(result)
        case "resume":
            handleResume(result)
        case "dispose":
            teardown()
            emitState("disposed")
            result(nil)
        case "reconfigure":
            teardown()
            handleStart(call, result)
        case "flushPlayback":
            playbackEngine?.flush()
            result(nil)
        case "play":
            let args = call.arguments as? [String: Any] ?? [:]
            if let pcm = args["pcm"] as? FlutterStandardTypedData {
                playbackEngine?.write(pcm: pcm.data)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "pcm bytes required", details: nil))
            }
        case "setRoute":
            result(nil) // iOS handles routing automatically in voiceChat mode
        case "getDiagnostics":
            handleGetDiagnostics(result)
        case "checkPermission":
            let args = call.arguments as? [String: Any] ?? [:]
            let type = args["type"] as? String ?? "microphone"
            result(permissionManager.check(type))
        case "requestPermission":
            let args = call.arguments as? [String: Any] ?? [:]
            let type = args["type"] as? String ?? "microphone"
            permissionManager.request(type, result: result)
        case "openAppSettings":
            permissionManager.openSettings()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleStart(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        let sampleRate = args["sampleRate"] as? Int ?? 16000
        let bufferMs = args["bufferDurationMs"] as? Int ?? 20
        let speakerMode = args["speakerMode"] as? Bool ?? true
        let preferredRouteName = args["preferredRoute"] as? String ?? "speaker"

        // Double-start guard: teardown any existing engine before re-creating.
        if currentState == "running" || currentState == "paused" || currentState == "starting" {
            teardown()
        }

        emitState("starting")
        lastConfig = args

        let session = AudioSessionManager()
        session.onInterruption = { [weak self] ended, shouldResume in
            if ended {
                if shouldResume {
                    self?.handleResume { _ in }
                }
            } else {
                self?.handlePause { _ in }
            }
        }
        session.onMediaServicesReset = { [weak self] in
            // Media services were reset — must fully recreate the engine.
            self?.teardown()
            if let config = self?.lastConfig {
                self?.handleStart(FlutterMethodCall(methodName: "start", arguments: config)) { _ in }
            }
        }
        do {
            try session.activate(sampleRate: Double(sampleRate),
                                 bufferDurationMs: bufferMs,
                                 speakerMode: speakerMode)
        } catch {
            teardown()
            emitState("error")
            publisher?.emitError(code: "START_FAILED", message: error.localizedDescription)
            result(FlutterError(code: "START_FAILED", message: error.localizedDescription, details: nil))
            return
        }
        sessionManager = session

        let pub = FramePublisher()
        pub.setEventSink(eventsSink)
        pub.setCaptureSink(captureSink)
        publisher = pub

        let route = AudioRouteManager()
        route.onRouteChanged = { [weak self] newRoute in
            guard let self = self else { return }
            self.publisher?.emitRouteChanged(previous: self.routeManager?.currentRoute.rawValue ?? "speaker",
                                             current: newRoute.rawValue)
        }
        route.start()
        routeManager = route

        let config = PipelineConfig.fromMap(args)
        let metadataStage = MetadataStage()
        let frameSize = sampleRate * bufferMs / 1000 * 2 // 16-bit mono
        let pipe = PipelineRunner(config: config, stages: [metadataStage], publisher: pub,
                                   frameSizeBytes: frameSize, sampleRate: sampleRate)
        pipe.initialize()
        pipe.start()
        pipeline = pipe

        let capture = AudioCaptureEngine()
        capture.targetSampleRate = Double(sampleRate)
        capture.onBuffer = { [weak pipe] buffer, _ in
            pipe?.handleBuffer(buffer)
        }
        do {
            try capture.start()
        } catch {
            teardown()
            emitState("error")
            publisher?.emitError(code: "START_FAILED", message: error.localizedDescription)
            result(FlutterError(code: "START_FAILED", message: error.localizedDescription, details: nil))
            return
        }
        captureEngine = capture

        // Playback shares the capture engine's AVAudioEngine so AEC can
        // correlate the playback reference with the capture input.
        let playback = AudioPlaybackEngine(engine: capture.engine)
        try? playback.start()
        playbackEngine = playback

        startTimeMs = Date().timeIntervalSince1970 * 1000
        emitState("running")
        result(nil)
    }

    private func handlePause(_ result: @escaping FlutterResult) {
        // Pause must stop capture — otherwise frames keep flowing while
        // the engine is supposedly suspended.
        pipeline?.stop()
        captureEngine?.stop()
        playbackEngine?.stop()
        emitState("paused")
        result(nil)
    }

    private func handleResume(_ result: @escaping FlutterResult) {
        // Resume must restart capture — the tap was removed during pause.
        playbackEngine?.start()
        pipeline?.start()
        do {
            try captureEngine?.start()
        } catch {
            emitState("error")
            publisher?.emitError(code: "RESUME_FAILED", message: error.localizedDescription)
            result(FlutterError(code: "RESUME_FAILED", message: error.localizedDescription, details: nil))
            return
        }
        emitState("running")
        result(nil)
    }

    private func handleGetDiagnostics(_ result: @escaping FlutterResult) {
        let session = AVAudioSession.sharedInstance()
        let uptimeMs = startTimeMs > 0 ? Int(Date().timeIntervalSince1970 * 1000 - startTimeMs) : 0

        // Real effect status — query the session for voice processing availability.
        let voiceProcessingAvailable: Bool
        if #available(iOS 13.0, *) {
            voiceProcessingAvailable = session.isInputAvailable
        } else {
            voiceProcessingAvailable = false
        }

        result([
            "deviceModel": UIDevice.current.model,
            "osVersion": UIDevice.current.systemVersion,
            "audioSessionId": 0,
            "effectStatus": [
                "aec": ["supported": voiceProcessingAvailable, "available": voiceProcessingAvailable, "active": currentState == "running"],
                "agc": ["supported": voiceProcessingAvailable, "available": voiceProcessingAvailable, "active": currentState == "running"],
                "noiseSuppression": ["supported": voiceProcessingAvailable, "available": voiceProcessingAvailable, "active": currentState == "running"],
            ],
            "currentRoute": routeManager?.currentRoute.rawValue ?? "speaker",
            "availableRoutes": (routeManager?.getAvailableRoutes() ?? [.speaker]).map(\.rawValue),
            "captureFrameCount": 0,
            "playbackFrameCount": 0,
            "uptimeMs": uptimeMs,
            "scoDeviceConnected": session.currentRoute.outputs.contains { $0.portType == .bluetoothHFP },
        ])
    }

    private func teardown() {
        pipeline?.stop()
        pipeline?.dispose()
        pipeline = nil
        captureEngine?.stop()
        captureEngine = nil
        playbackEngine?.stop()
        playbackEngine = nil
        routeManager?.stop()
        routeManager = nil
        sessionManager?.deactivate()
        sessionManager = nil
        publisher = nil
        startTimeMs = 0
    }

    private func emitState(_ newState: String) {
        publisher?.emitStateChange(previous: currentState, current: newState)
        currentState = newState
    }
}

// MARK: - Stream Handlers

extension BoatPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventsSink = events
        publisher?.setEventSink(events)
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventsSink = nil
        publisher?.setEventSink(nil)
        return nil
    }
}

private final class CaptureStreamHandler: NSObject, FlutterStreamHandler {
    private let onSink: (FlutterEventSink?) -> Void

    init(onSink: @escaping (FlutterEventSink?) -> Void) {
        self.onSink = onSink
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        onSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        onSink(nil)
        return nil
    }
}

File: src/android/src/main/kotlin/com/circuids/boat/BoatPlugin.kt
package com.circuids.boat

import android.app.Activity
import android.content.Context
import com.circuids.boat.audio.*
import com.circuids.boat.permission.PermissionManager
import com.circuids.boat.pipeline.*
import com.circuids.boat.pipeline.stages.MetadataStage
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class BoatPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var methodsChannel: MethodChannel
    private lateinit var eventsChannel: EventChannel
    private lateinit var captureChannel: EventChannel

    @Volatile
    private var eventsSink: EventChannel.EventSink? = null
    @Volatile
    private var captureSink: EventChannel.EventSink? = null

    private var sessionManager: AudioSessionManager? = null
    private var captureEngine: AudioCaptureEngine? = null
    private var playbackEngine: AudioPlaybackEngine? = null
    private var effectsManager: EffectsManager? = null
    private var routeManager: AudioRouteManager? = null
    private var pipeline: PipelineRunner? = null
    private var publisher: FramePublisher? = null

    private var currentState = "idle"
    private var startTimeMs = 0L
    private lateinit var appContext: Context
    private var permissionManager: PermissionManager? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        val messenger = binding.binaryMessenger

        methodsChannel = MethodChannel(messenger, "com.circuids.boat/methods")
        methodsChannel.setMethodCallHandler(this)

        eventsChannel = EventChannel(messenger, "com.circuids.boat/events")
        eventsChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventsSink = events
            }
            override fun onCancel(arguments: Any?) {
                eventsSink = null
            }
        })

        captureChannel = EventChannel(messenger, "com.circuids.boat/capture")
        captureChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                captureSink = events
            }
            override fun onCancel(arguments: Any?) {
                captureSink = null
            }
        })

        permissionManager = PermissionManager(appContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        teardown()
        methodsChannel.setMethodCallHandler(null)
        permissionManager = null
    }

    // ── ActivityAware ──

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        permissionManager?.setActivity(binding.activity)
        binding.addRequestPermissionsResultListener(
            permissionManager ?: return
        )
    }

    override fun onDetachedFromActivityForConfigChanges() {
        permissionManager?.setActivity(null)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        permissionManager?.setActivity(binding.activity)
        binding.addRequestPermissionsResultListener(
            permissionManager ?: return
        )
    }

    override fun onDetachedFromActivity() {
        permissionManager?.setActivity(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "start" -> handleStart(call, result)
            "stop" -> handleStop(result)
            "pause" -> handlePause(result)
            "resume" -> handleResume(result)
            "dispose" -> handleDispose(result)
            "reconfigure" -> handleReconfigure(call, result)
            "flushPlayback" -> handleFlushPlayback(result)
            "play" -> handlePlay(call, result)
            "setRoute" -> handleSetRoute(call, result)
            "getDiagnostics" -> handleGetDiagnostics(result)
            "checkPermission" -> {
                val type = call.argument<String>("type") ?: "microphone"
                result.success(permissionManager?.check(type) ?: "denied")
            }
            "requestPermission" -> {
                val type = call.argument<String>("type") ?: "microphone"
                permissionManager?.request(type, result) ?: result.success("denied")
            }
            "openAppSettings" -> {
                permissionManager?.openSettings()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun handleStart(call: MethodCall, result: Result) {
        try {
            // Double-start guard: teardown any existing engine before
            // re-creating to prevent resource leaks.
            if (currentState == "running" || currentState == "paused" || currentState == "starting") {
                teardown()
            }

            val args = call.arguments as? Map<*, *> ?: emptyMap<String, Any>()
            val sampleRate = (args["sampleRate"] as? Number)?.toInt() ?: 16000
            val bufferMs = (args["bufferDurationMs"] as? Number)?.toInt() ?: 20
            val enableAec = args["aec"] as? Boolean ?: true
            val enableAgc = args["agc"] as? Boolean ?: true
            val enableNs = args["noiseSuppression"] as? Boolean ?: true
            val speakerMode = args["speakerMode"] as? Boolean ?: true
            val preferredRouteName = args["preferredRoute"] as? String ?: "speaker"

            emitState("starting")

            val session = AudioSessionManager(appContext)
            session.onFocusLost = {
                publisher?.emitWarning("AUDIO_FOCUS_LOST", "Audio focus lost — pausing capture")
                pipeline?.stop()
                playbackEngine?.stop()
                emitState("paused")
            }
            session.activate()
            sessionManager = session

            val capture = AudioCaptureEngine(sampleRate, bufferMs)
            val sessionId = capture.create()
            captureEngine = capture

            val effects = EffectsManager()
            effects.attach(sessionId, enableAec, enableAgc, enableNs)
            effectsManager = effects

            val pub = FramePublisher(
                captureSink = { captureSink },
                eventSink = { eventsSink },
            )
            publisher = pub

            // Route must be applied BEFORE creating AudioTrack — some devices
            // bind the track to the active communication device at creation time.
            val route = AudioRouteManager(appContext, session.audioManagerRef) { newRoute ->
                publisher?.emitRouteChanged(currentRouteName(), newRoute.channelName)
            }
            route.start()
            val preferredRoute = BoatAudioRoute.fromChannelName(preferredRouteName)
            route.setPolicy(RoutePolicy(speakerMode, preferredRoute))
            route.applyPolicy(force = true)
            routeManager = route

            val playback = AudioPlaybackEngine(sampleRate, sessionId)
            playback.create()
            playback.start()
            playbackEngine = playback

            val config = PipelineConfig.fromMap(args.mapKeys { it.key.toString() })
            val metadataStage = MetadataStage { route.currentRoute.channelName }
            // CapturePipeline owns the capture thread + pre-allocated frame
            val pipe = PipelineRunner(config, listOf(metadataStage), pub, capture, sampleRate)
            pipe.initialize()
            pipe.start()
            pipeline = pipe

            startTimeMs = System.currentTimeMillis()
            emitState("running")
            result.success(null)
        } catch (e: Exception) {
            // Partial-failure cleanup: tear down any resources that were
            // allocated before the exception so we don't leak.
            teardown()
            emitState("error")
            publisher?.emitError("START_FAILED", e.message ?: "Unknown start failure")
            result.error("START_FAILED", e.message, null)
        }
    }

    private fun handleStop(result: Result) {
        teardown()
        emitState("idle")
        result.success(null)
    }

    private fun handlePause(result: Result) {
        pipeline?.stop()
        playbackEngine?.stop()
        emitState("paused")
        result.success(null)
    }

    private fun handleResume(result: Result) {
        pipeline?.start()
        playbackEngine?.start()
        emitState("running")
        result.success(null)
    }

    private fun handleDispose(result: Result) {
        teardown()
        emitState("disposed")
        result.success(null)
    }

    private fun handleReconfigure(call: MethodCall, result: Result) {
        // Stop and restart with new config
        teardown()
        handleStart(call, result)
    }

    private fun handleFlushPlayback(result: Result) {
        playbackEngine?.flush()
        result.success(null)
    }

    private fun handlePlay(call: MethodCall, result: Result) {
        val pcm = call.argument<ByteArray>("pcm")
        if (pcm == null) {
            result.error("INVALID_ARGUMENT", "pcm bytes required", null)
            return
        }
        playbackEngine?.write(pcm)
        result.success(null)
    }

    private fun handleSetRoute(call: MethodCall, result: Result) {
        val routeName = call.argument<String>("route") ?: "speaker"
        val route = BoatAudioRoute.fromChannelName(routeName)
        routeManager?.setRoute(route)
        result.success(null)
    }

    private fun handleGetDiagnostics(result: Result) {
        val effects = effectsManager?.getStatus() ?: emptyMap()
        val effectMap = effects.mapValues { (_, state) ->
            mapOf("supported" to state.supported, "available" to state.available, "active" to state.active)
        }
        val uptimeMs = if (startTimeMs > 0) System.currentTimeMillis() - startTimeMs else 0
        result.success(
            mapOf(
                "deviceModel" to android.os.Build.MODEL,
                "osVersion" to android.os.Build.VERSION.RELEASE,
                "audioSessionId" to (captureEngine?.audioSessionId ?: -1),
                "effectStatus" to effectMap,
                "currentRoute" to currentRouteName(),
                "availableRoutes" to (routeManager?.getAvailableRoutes()?.map { it.channelName } ?: listOf("speaker")),
                "captureFrameCount" to (pipeline?.captureFrameCount ?: 0),
                "playbackFrameCount" to (playbackEngine?.playbackFrameCount ?: 0),
                "uptimeMs" to uptimeMs,
                "scoDeviceConnected" to (routeManager?.isScoDeviceConnected() ?: false),
            )
        )
    }

    private fun teardown() {
        pipeline?.stop()
        pipeline?.dispose()
        pipeline = null
        captureEngine?.release()
        captureEngine = null
        playbackEngine?.release()
        playbackEngine = null
        effectsManager?.release()
        effectsManager = null
        routeManager?.stop()
        routeManager = null
        sessionManager?.deactivate()
        sessionManager = null
        publisher = null
        startTimeMs = 0
    }

    private fun emitState(newState: String) {
        publisher?.emitStateChange(currentState, newState)
        currentState = newState
    }

    private fun currentRouteName(): String =
        routeManager?.currentRoute?.channelName ?: "speaker"
}
